main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2021 MobilityData IO
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.validator;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.mobilitydata.gtfsvalidator.annotation.GtfsValidator;
+import org.mobilitydata.gtfsvalidator.annotation.Inject;
+import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
+import org.mobilitydata.gtfsvalidator.notice.StartAndEndDateOutOfOrderNotice;
+import org.mobilitydata.gtfsvalidator.notice.StopTooFarFromTripShapeNotice;
+import org.mobilitydata.gtfsvalidator.table.GtfsShapeTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsStop;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTimeTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsTrip;
+import org.mobilitydata.gtfsvalidator.table.GtfsTripTableContainer;
+import org.mobilitydata.gtfsvalidator.util.GeospatialUtil;
+
+/**
+ * Validates: a {@link GtfsStop} is within a distance threshold for a trip shape.
+ *
+ * <p>Generated notice: {@link StartAndEndDateOutOfOrderNotice}.
+ */
+@GtfsValidator
+public class StopTooFarFromTripShapeValidator extends FileValidator {
+  @Inject GtfsStopTimeTableContainer stopTimeTable;
+  @Inject GtfsTripTableContainer tripTable;
+  @Inject GtfsShapeTableContainer shapeTable;
+  @Inject GtfsStopTableContainer stopTable;
+
+  @Override
+  public void validate(NoticeContainer noticeContainer) {
+    List<StopTooFarFromTripShapeNotice> notices = new ArrayList<>();
+
+    // Cache for previously tested shape_id and stop_id pairs - we don't need to test them more than
+    // once
+    final Set<String> testedCache = new HashSet<>();
+
+    stopTimeTable
+        .byTripIdMap()
+        .forEach(
+            (tripId, tripStopTimes) -> {
```

Comment:
Here you are checking every trip as many times as you have it in stop_times.txt. That is a bug in the algorithm and a serious performance issue.

What was the size of the largest feed that you used for testing? I used a medium-size feed with 1 million stop_times.txt and this code runs for several minutes on it. If we add a large enough feed to the test set, we can catch such bugs earlier.

Reply:
Another way to find this bug earlier was to see that you are producing the same notice multiple times. Can you add unit tests for that?

Reply:
> If we add a large enough feed to the test set, we can catch such bugs earlier.

I think that might explains what happens with Buenos Aires
![image](https://user-images.githubusercontent.com/416651/107685651-3d924600-6c72-11eb-9133-f9872a4a299c.png)
![image](https://user-images.githubusercontent.com/416651/107685725-53077000-6c72-11eb-83c5-f8734f531d78.png)
https://github.com/MobilityData/gtfs-validator/runs/1881156285?check_suite_focus=true

Reply:
This piece of code origins from: https://github.com/MobilityData/gtfs-validator/pull/340. 
@barbeau could you mind taking a look at this please?

Reply:
I need to look at this a little closer, but I think this is a translation issue from the original implementation due to the differences in data structures between v1 and v2. 

In the original implementation of v1, the in-memory GTFS data repository held a Map data structure where each trip_id appeared once as the key and the value was a SortedMap of StopTimes for that trip_id. This rule looped through each of those entries and therefore only processed each trip once.

But the similarly named variable in v2 actually contains the entire contents of stop_times.txt. So it looks like this implementation is fetching all StopTimes and then getting the trip_id for each StopTime and then getting all the StopTimes for that trip_id, which blows up the processing time significantly.

I'll take a closer look tomorrow.

Reply:
@aababilov Are you working on re-implementing this with a separate spatial library? Or should MobilityData work on fixing the implementation issues with this rule using Spatial4j?

Reply:
> This piece of code origins from: #340.

It actually does not origin from the Sean's commit. Sean wrote another logic that iterated on each trip only once. The new code, however, iterates on each pair of <trip_id, GtfsStopTime>. IDEA can spell out the types for you:

```
    stopTimeTable
        .byTripIdMap()
        .forEach(
            (String tripId, GtfsStopTime tripStopTimes) -> {
                ....
```

Also, the new code does not use the `tripStopTimes` variable - instead, it calls `stopTimeTable.byTripId(tripId)` that really returns `List<GtfsStopTime>`. If you use your `tripStopTimes` in that place, that won't compile and you would catch this bug.

Reply:
Overall, I think I can suggest you to spell out the types and also prefer plain cycles to lambdas. That will help to avoid such kind of bugs in the future.

Reply:
> @aababilov Are you working on re-implementing this with a separate spatial library? Or should MobilityData work on fixing the implementation issues with this rule using Spatial4j?

This validation rule is not the top priority for me right now. What I did was to temporarily remove that code at Google.

I think that we should put as a priority that our MobilityData upstream code works as expected for a large set of feeds (such as several thousands that we have at Google). I do not think that we need to rush implementing as many validations as possible without proper testing since that puts too much technical debt for the future maintainers.

I am always testing my code on all Google feeds and I check the results, so that gives a certain confidence. I see that MobilityData is not actually testing the code on a considerable feed set. Even more, the test results may be misinterpreted (as it happened for the Montreal feed), so that reduces positive impact of such testing.

@barbeau @lionel-nj Are you able to setup a reliable testing procedure on your side and run the Validator on a large set of feeds (hundreds or better thousands)? Without such testing, we always risk to accumulate bugs in the upstream code. May we hold on all code contributions from MobilityData before such testing procedure is established?

Reply:
@aababilov could you provide the url of the dataset you were looking at please?

Reply:
> I used a medium-size feed with 1 million stop_times.txt and this code runs for several minutes on it.


@aababilov could you provide the url of the dataset you were looking at please?